File: mobile/modules/native-storage/ios/NativeStorageModule.swift
import ExpoModulesCore
import Foundation

/// Read-only bridge over `UserDefaults.standard` — the same store the native iOS app writes.
/// Never deletes or overwrites keys; Expo hydrates AsyncStorage from this snapshot once.
public class NativeStorageModule: Module {
  public func definition() -> ModuleDefinition {
    Name("NativeStorage")

    Constants([
      "isAvailable": true,
      "platform": "ios",
    ])

    AsyncFunction("readSnapshot") { () -> [String: Any] in
      let defaults = UserDefaults.standard
      var blobs: [String: String] = [:]
      var prefs: [String: Any] = [:]

      for key in Self.blobKeys {
        if let string = Self.stringValue(defaults, key: key) {
          blobs[key] = string
        }
      }

      for key in Self.prefKeys {
        guard defaults.object(forKey: key) != nil else { continue }
        if let value = Self.jsPref(defaults.object(forKey: key)) {
          prefs[key] = value
        }
      }

      return [
        "available": true,
        "platform": "ios",
        "blobs": blobs,
        "prefs": prefs,
        "foodImagesDirectory": Self.foodImagesDirectoryPath() as Any,
      ]
    }

    AsyncFunction("copyFoodImages") { (destination: String) -> Int in
      Self.copyFoodImages(to: destination)
    }
  }

  /// JSON blobs persisted as `Data` (PersistedBlobGuard) or, rarely, as a UTF-8 string.
  private static let blobKeys: [String] = [
    "foodEntries",
    "favoriteFoodEntries",
    "favorites",
    "waterEntries",
    "fastingSessions",
    "weightEntries",
    "bodyFatEntries",
    "userProfile",
    "coachChatHistory",
    "fudai.workouts.diary.state.v1",
    "workoutDiaryStateV1",
    "adaptiveGoalsPreviousTargets",
  ]

  /// Property names match `Preferences` / iOS `@AppStorage` keys, plus Android aliases
  /// so one snapshot shape works on both platforms.
  private static let prefKeys: [String] = [
    "hasCompletedOnboarding",
    "appearanceMode",
    "appThemeColor",
    "heightUnit",
    "weightUnit",
    "useMetric",
    "weekStartsOnMonday",
    "homeTopNutrients",
    "foodLogSortOrder",
    "waterTrackingEnabled",
    "waterDailyGoalMl",
    "waterUnit",
    "waterReminderEnabled",
    "waterReminderHour",
    "waterReminderMinute",
    "fastingTrackingEnabled",
    "fastingDefaultGoalMinutes",
    "fastingGoalNotificationEnabled",
    "notificationsEnabled",
    "breakfastReminderEnabled",
    "breakfastReminderHour",
    "breakfastReminderMinute",
    "lunchReminderEnabled",
    "lunchReminderHour",
    "lunchReminderMinute",
    "dinnerReminderEnabled",
    "dinnerReminderHour",
    "dinnerReminderMinute",
    "healthKitEnabled",
    "healthConnectEnabled",
    "adaptiveGoalsEnabled",
    "adaptiveGoalsPreviousTargets",
    "adaptiveGoalsLastCheckDay",
    "selectedSpeechProvider",
    "onboardingPlanEdited",
    "aiAccessMode",
    "aiConsentGiven",
    "aiAnalysisConsentGiven",
    "acceptedTermsAndPrivacy",
    "selectedAIProvider",
    "selectedAIModel",
    "separateTextProviderEnabled",
    "selectedTextAIProvider",
    "selectedTextAIModel",
    "aiUserContext",
    "userContext",
    "aiFallbackEnabled",
    "selectedFallbackAIProvider",
    "selectedFallbackAIModel",
    "aiMaxResponseTokens",
    "maxResponseTokens",
    "aiRequestTimeoutSeconds",
  ]

  private static func stringValue(_ defaults: UserDefaults, key: String) -> String? {
    if let data = defaults.data(forKey: key), let string = String(data: data, encoding: .utf8), !string.isEmpty {
      return string
    }
    if let string = defaults.string(forKey: key), !string.isEmpty {
      return string
    }
    return nil
  }

  private static func jsPref(_ value: Any?) -> Any? {
    switch value {
    case let flag as Bool:
      return flag
    case let number as Int:
      return number
    case let number as Double:
      return number
    case let number as Float:
      return Double(number)
    case let string as String:
      return string
    case let data as Data:
      return String(data: data, encoding: .utf8)
    case let number as NSNumber:
      if CFGetTypeID(number) == CFBooleanGetTypeID() {
        return number.boolValue
      }
      if CFNumberIsFloatType(number) {
        return number.doubleValue
      }
      return number.intValue
    default:
      return nil
    }
  }

  /// Native `FoodImageStore` writes JPEGs under Application Support, not Documents.
  private static func foodImagesDirectoryPath() -> String? {
    foodImagesSourceDirectories().first?.path
  }

  private static func fileURL(_ path: String) -> URL {
    if path.hasPrefix("file:"), let url = URL(string: path) {
      return url
    }
    return URL(fileURLWithPath: path)
  }

  private static func foodImagesSourceDirectories() -> [URL] {
    var urls: [URL] = []
    if let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first {
      urls.append(support.appendingPathComponent("fudai-food-images", isDirectory: true))
    }
    if let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {
      urls.append(documents.appendingPathComponent("fudai-food-images", isDirectory: true))
    }
    return urls
  }

  private static func copyFoodImages(to destination: String) -> Int {
    let dest = fileURL(destination)
    try? FileManager.default.createDirectory(at: dest, withIntermediateDirectories: true)
    var copied = 0
    for source in foodImagesSourceDirectories() {
      copied += copyJPEGs(from: source, to: dest)
    }
    return copied
  }

  private static func copyJPEGs(from source: URL, to dest: URL) -> Int {
    let fm = FileManager.default
    guard fm.fileExists(atPath: source.path),
          let items = try? fm.contentsOfDirectory(at: source, includingPropertiesForKeys: nil) else {
      return 0
    }
    if (try? source.resourceValues(forKeys: [.canonicalPathKey]).canonicalPath)
        == (try? dest.resourceValues(forKeys: [.canonicalPathKey]).canonicalPath) {
      return items.filter { isImage($0) }.count
    }
    var copied = 0
    for file in items where isImage(file) {
      let target = dest.appendingPathComponent(file.lastPathComponent)
      if fm.fileExists(atPath: target.path) {
        copied += 1
        continue
      }
      do {
        try fm.copyItem(at: file, to: target)
        copied += 1
      } catch {
        continue
      }
    }
    return copied
  }

  private static func isImage(_ url: URL) -> Bool {
    ["jpg", "jpeg", "png", "webp"].contains(url.pathExtension.lowercased())
  }
}

File: mobile/modules/native-storage/android/src/main/java/expo/modules/nativestorage/NativeStorageModule.kt
package expo.modules.nativestorage

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.PreferenceDataStoreFactory
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.preferencesDataStoreFile
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext

/**
 * Read-only bridge over the native Android Preferences DataStore `fudai_prefs`.
 * Same file [FudaiDataStore] / [PreferencesStore] write — never a second store, never a wipe.
 */
class NativeStorageModule : Module() {
  private val storeMutex = Mutex()
  @Volatile private var store: DataStore<Preferences>? = null

  override fun definition() = ModuleDefinition {
    Name("NativeStorage")

    Constants {
      mapOf(
        "isAvailable" to true,
        "platform" to "android",
      )
    }

    AsyncFunction("readSnapshot") Coroutine {
      readSnapshotOrEmpty()
    }

    AsyncFunction("copyFoodImages") Coroutine { destination: String ->
      withContext(Dispatchers.IO) {
        copyFoodImages(destination)
      }
    }
  }

  private suspend fun readSnapshotOrEmpty(): Map<String, Any?> = withContext(Dispatchers.IO) {
    val dataStore = dataStoreOrNull()
    if (dataStore == null) {
      return@withContext mapOf(
        "available" to false,
        "platform" to "android",
        "blobs" to emptyMap<String, String>(),
        "prefs" to emptyMap<String, Any>(),
        "foodImagesDirectory" to foodImagesDirectoryPath(),
      )
    }

    val prefs = try {
      dataStore.data.first()
    } catch (_: Exception) {
      // Leave the on-disk file untouched — do not install a replace-on-corrupt handler.
      return@withContext mapOf(
        "available" to false,
        "platform" to "android",
        "blobs" to emptyMap<String, String>(),
        "prefs" to emptyMap<String, Any>(),
        "foodImagesDirectory" to foodImagesDirectoryPath(),
      )
    }

    val blobs = linkedMapOf<String, String>()
    val scalars = linkedMapOf<String, Any>()
    for ((key, value) in prefs.asMap()) {
      val name = key.name
      when (value) {
        is String -> {
          if (name in blobKeys || looksLikeJson(value)) blobs[name] = value
          else scalars[name] = value
        }
        is Boolean -> scalars[name] = value
        is Int -> scalars[name] = value
        is Long -> scalars[name] = value.toDouble()
        is Float -> scalars[name] = value.toDouble()
        is Double -> scalars[name] = value
        is Set<*> -> {
          val strings = value.mapNotNull { it as? String }
          blobs[name] = org.json.JSONArray(strings).toString()
        }
      }
    }

    mapOf(
      "available" to true,
      "platform" to "android",
      "blobs" to blobs,
      "prefs" to scalars,
      "foodImagesDirectory" to foodImagesDirectoryPath(),
    )
  }

  private suspend fun dataStoreOrNull(): DataStore<Preferences>? {
    val context = appContext.reactContext ?: return null
    store?.let { return it }
    return storeMutex.withLock {
      store?.let { return@withLock it }
      val created = PreferenceDataStoreFactory.create(
        produceFile = { context.applicationContext.preferencesDataStoreFile(DATASTORE_NAME) },
      )
      store = created
      created
    }
  }

  private fun foodImagesDirectoryPath(): String? {
    val context = appContext.reactContext ?: return null
    return File(context.filesDir, FOOD_IMAGES_DIR).absolutePath
  }

  private fun copyFoodImages(destination: String): Int {
    val dest = fileFromPath(destination)
    dest.mkdirs()
    val sourcePath = foodImagesDirectoryPath() ?: return 0
    val source = File(sourcePath)
    if (!source.isDirectory) return 0
    if (source.canonicalPath == dest.canonicalPath) {
      return source.listFiles()?.count { isImage(it) } ?: 0
    }
    var copied = 0
    source.listFiles()?.forEach { file ->
      if (!isImage(file)) return@forEach
      val target = File(dest, file.name)
      if (target.exists() || runCatching { file.copyTo(target, overwrite = false) }.isSuccess) {
        copied += 1
      }
    }
    return copied
  }

  private fun fileFromPath(path: String): File =
    if (path.startsWith("file:")) File(android.net.Uri.parse(path).path ?: path) else File(path)

  private fun isImage(file: File): Boolean {
    val ext = file.extension.lowercase()
    return ext == "jpg" || ext == "jpeg" || ext == "png" || ext == "webp"
  }

  private companion object {
    const val DATASTORE_NAME = "fudai_prefs"
    const val FOOD_IMAGES_DIR = "fudai-food-images"

    val blobKeys = setOf(
      "foodEntries",
      "favoriteFoodEntries",
      "favorites",
      "waterEntries",
      "fastingSessions",
      "weightEntries",
      "bodyFatEntries",
      "userProfile",
      "coachChatHistory",
      "workoutDiaryStateV1",
      "fudai.workouts.diary.state.v1",
      "adaptiveGoalsPreviousTargets",
    )

    fun looksLikeJson(value: String): Boolean {
      val trimmed = value.trim()
      return (trimmed.startsWith("{") && trimmed.endsWith("}")) ||
        (trimmed.startsWith("[") && trimmed.endsWith("]"))
    }
  }
}
